[core] Add support for dynamic chunks

Alxiice opened the pull request:
## Description

This PR goal is to change chunk system to be able to work with graphs where nodes can have no specific size set.

## Changes

**Compute system**
- New levels have been added on the computation : 
  - `extreme` (that targets higher specs that what intensive does)
  - `script` which is a mode for single-process simple process. It targets specific machines on the farm that have the ability to run multiple jobs in parallel so for simple jobs it should allow faster processing
- Delayed chunks creation
  - `node`
    - A `_chunksCreated` param has been added to check if the chunks have been correctly initialized
    - Now by default we use `resetChunks` when loading meshroom, that will create a list with a single chunk, and we can create chunks at any moment with `_createChunks`.
    - A new `NodeStatusData` has been added that tracks a `nodeStatus` file which is similar to the chunk statuses files but is specific to the node. This node status is used to get the cached range parametrization of the node so that we can retrieve it when possible with `_createChunksFromCache`.
  - `taskManager`
    - The `TaskThread` and `TaskManager` have been modified to launch the chunk creation when the node compute starts

**Submitters**
- Submitters have been moved to another package (TBD where...)
- The `BaseSubmitter` API have been updated to allow more flexibility. A new `BaseSubmittedJob` exists and is used to track jobs that have been created. The goal is to use this as an interface to call actions on it (stop/pause/restart...)
- A new `bin/meshroom_createChunks` script have been created, and handles the chunk creation and additional chunk tasks spooling
  - Calls the chunk creation
  - Checks if we can spool additioanl tasks
  - If so, execute queue tasks that will compute the chunks
  - If not, execute the chunks serially on the current process

Changes to the tractor API have been implemented here https://github.com/meshroomHub/mrSubmitters/pull/1

## Examples

![Peek 2025-10-09 19-05](https://github.com/user-attachments/assets/21a91f28-df25-4705-8a34-96f963c69171)

![ezgif-1a582893f45983](https://github.com/user-attachments/assets/17a36313-0b6d-436a-aea2-6be55c5180b2)

Alxiice committed:
[core] node/taskManager: create _chunksCreated to delay chunk creation, create a NodeStatusData

Alxiice committed:
[node] Add licenses list on node desc to provide a source where we can fetch licenses required for the submitter

Alxiice committed:
[core] computation : update computation levels

Alxiice committed:
[bin] Add createChunks script

Alxiice committed:
[core] Update submitter API

Alxiice committed:
[submitter] Fix SubmitterOptionsEnum.ALL mode on py 3.9

Alxiice committed:
[qml] Fix anchor issue when chunks are emptied

Alxiice committed:
[core] Node : add defaultStatus in _createChunks

Alxiice committed:
[core] Start updating taskmanager and submitter for new chunk process

Alxiice committed:
[core] First implementation to kill submitted tasks

Alxiice committed:
[core] graph : Manage nodeStatus file monitoring

Alxiice committed:
[code] submitter : fix issues in dynamic chunks & submitting

Alxiice committed:
[code] graph : Better management of statuses after task/job actions

Alxiice committed:
[core] Fix issues on missing chunks fro sfm node & node chunk indicator

Alxiice committed:
[core] submitter : retrieve job on node update + fix some ui issues

cbentejac commented on meshroom/core/node.py:
```diff
@@ -164,57 +313,39 @@ def initLocalSubmit(self):
     def elapsedTimeStr(self):
         return str(datetime.timedelta(seconds=self.elapsedTime))
-
+    
```
Trailing whitespaces

cbentejac commented on bin/meshroom_createChunks:
```diff
@@ -0,0 +1,145 @@
+                f"\"{chunk.status.status.name}\". See file: \"{chunk.statusFile}\". " \
+                f"ExecMode: {chunk.status.execMode.name}, computeSessionUid: {chunk.status.computeSessionUid}, " \
+                f"submitterSessionUid: {node._nodeStatus.submitterSessionUid}")
+            # sys.exit(-1)
```
To remove if it's not going to be used

cbentejac commented on meshroom/core/node.py:
```diff
@@ -2230,3 +2599,5 @@ def upgrade(self):
     canUpgrade = Property(bool, canUpgrade.fget, constant=True)
     issueDetails = Property(str, issueDetails.fget, constant=True)
 
+    issueDetails = Property(str, issueDetails.fget, constant=True)
```
```suggestion
```

This seems to be a duplication from line 2600, it should be removed.

cbentejac commented on meshroom/ui/qml/GraphEditor/TaskManager.qml:
```diff
@@ -51,222 +62,430 @@ Item {
+            // Job toolbar
+            Rectangle {
+                // Layout.preferredWidth: childrenRect.width
+                // Layout.preferredHeight: childrenRect.height
```
```suggestion
```

cbentejac commented on meshroom/ui/qml/GraphEditor/TaskManager.qml:
```diff
@@ -51,222 +62,430 @@ Item {
+                            id: chunkDelegate
+                            // width: ListView.view.model 
+                            //     ? ListView.view.width / ListView.view.model.count - chunkList.spacing
+                            //     : 0
```
```suggestion
```

cbentejac commented on meshroom/ui/qml/GraphEditor/TaskManager.qml:
```diff
@@ -51,222 +62,430 @@ Item {
-                                selectNode(chunkList.node)
+                            background: Rectangle {
+                                color: Colors.darkpurple
+                                // color: Colors.getNodeColor(chunkList.node, {"NONE": Colors.darkpurple})
```
```suggestion
```

Alxiice replied:
Here the darkpurple was a debug tool. Instead I will use the commented line, so that for scheduled nodes without chunks we see the submission status

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+        else:
+            self.cancelNodeComputation(node)
+            node.stopComputation()
+            # self._taskManager.removeNode(node, processList=True)
```
```suggestion
```

cbentejac commented on meshroom/core/desc/computation.py:
```diff
@@ -1,13 +1,15 @@
     NONE = 0
     NORMAL = 1
     INTENSIVE = 2
+    EXTREME = 2
```
Shouldn't this be:
```suggestion
    EXTREME = 3
```
as used in https://github.com/meshroomHub/mrSubmitters/pull/1/files#diff-9dd253882930191e431fa00981aa86f2b39630037daed18a8fc848eaf3766823R77?

Alxiice replied:
yes

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+            else:
+                self.parent().showMessage(f"Paused node {node.label} on farm")
+        elif not node.isExtern():
+            self.parent().showMessage(f"PauseJob is only available in local computation mode !", "warning")
```
```suggestion
            self.parent().showMessage(f"PauseJob is only available in external computation mode!", "warning")
```

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+            try:
+                job.resumeJob()
+            except Exception as e:
+                self.parent().showMessage(f"Failed to rsume node {node.label} on farm")
```
```suggestion
                self.parent().showMessage(f"Failed to resume node {node.label} on farm")
```

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+                job.resumeJob()
+            except Exception as e:
+                self.parent().showMessage(f"Failed to rsume node {node.label} on farm")
+                logging.warning(f"Error on resumeJob :\n{e}")
```
```suggestion
                logging.warning(f"Error on resumeJob:\n{e}")
```

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+                job.interruptJob()
+            except Exception as e:
+                self.parent().showMessage(f"Failed to interrupt node {node.label} on farm", "error")
+                logging.warning(f"Error on interruptJob :\n{e}")
```
```suggestion
                logging.warning(f"Error on interruptJob:\n{e}")
```

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+                job.resumeJob()
+            except Exception as e:
+                self.parent().showMessage(f"Failed to restart error tasks for node {node.label} on farm", "error")
+                logging.warning(f"Error on restartJobErrorTasks :\n{e}")
```
```suggestion
                logging.warning(f"Error on restartJobErrorTasks:\n{e}")
```

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+                self.parent().showMessage(f"Failed to restart error tasks for node {node.label} on farm", "error")
+                logging.warning(f"Error on restartJobErrorTasks :\n{e}")
+            else:
+                self.parent().showMessage(f"Restared error tasks for the node {node}")
```
```suggestion
                self.parent().showMessage(f"Restarted error tasks for the node {node}")
```

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+            for chunk in node._chunks:
+                if chunk._status.status == Status.RUNNING and not chunk.isExtern():
+                    chunk.stopProcess()
+            self.parent().showMessage(f"Cleared the task manager")
```
When entering this case, the task manager is correctly cleared but in the Graph Editor, nodes that follow the one that was interrupted still remain locked and in "SUBMITTED" status, and appear as though they will be computed (with the start/stop/restart button in the "stop" position).

For them to return to a "normal" state, it is necessary to do a "stop" from the Graph Editor.

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+            try:
+                job.pauseJob()
+            except Exception as e:
+                logging.warning(f"Error on pauseJob :\n{e}")
```
```suggestion
                logging.warning(f"Error on pauseJob:\n{e}")
```

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+                job.skipChunkTask(node, chunkIteration)
+            except Exception as e:
+                self.parent().showMessage(f"Failed to skip chunk {chunkIteration} of {node.label}", "error")
+                logging.warning(f"Error on skipTask :\n{e}")
```
```suggestion
                logging.warning(f"Error on skipTask:\n{e}")
```

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+                chunk.updateStatusFromCache()
+                chunk.upgradeStatusTo(Status.ERROR)
+                self.parent().showMessage(f"Failed to relaunch chunk {chunkIteration} of {node.label}", "error")
+                logging.warning(f"Error on restartTask :\n{e}")
```
```suggestion
                logging.warning(f"Error on restartTask:\n{e}")
```

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+                job.stopChunkTask(node, -1)
+            except Exception as e:
+                self.parent().showMessage(f"Failed to stop node {node.label}", "error")
+                logging.warning(f"Error on stopTask :\n{e}")
```
```suggestion
                logging.warning(f"Error on stopTask:\n{e}")
```

cbentejac commented on meshroom/ui/graph.py:
```diff
@@ -570,6 +591,217 @@ def cancelNodeComputation(self, node):
+                job.stopChunkTask(node, chunkIteration)
+            except Exception as e:
+                self.parent().showMessage(f"Failed to stop chunk {chunkIteration} of {node.label}", "error")
+                logging.warning(f"Error on stopTask :\n{e}")
```
```suggestion
                logging.warning(f"Error on stopTask:\n{e}")
```

cbentejac commented on meshroom/core/submitter.py:
```diff
@@ -1,23 +1,228 @@
+        else:
+            raise RuntimeError(f"Submitter {self.__class__.__name__} cannot interrupt the job")
+
+    def interruptJob(self):
+        """ This will interrupt the job (and kill running tasks) """
+        if self.submitterOptions.includes(SubmitterOptionsEnum.INTERRUPT_JOB):
+            raise NotImplementedError("'interruptJob' method must be implemented in subclasses")
+        else:
+            raise RuntimeError(f"Submitter {self.__class__.__name__} cannot interrupt the job")
```
This works well to interrupt the job (on the farm, everything is updated as expected and the running process is killed), but there seems to be missing updates on the UI-side:

<img width="1419" height="310" alt="Image" src="https://github.com/user-attachments/assets/f00031ee-7615-4766-a3e1-3a431f44b945" />

All the submitted tasks have been stopped, and the running one still appears as running (both in the task manager and in the Graph Editor) although it has been successfully stopped on the farm. Refreshing the nodes' status has no effect.

Alxiice replied:
It should be better now

cbentejac commented on meshroom/ui/qml/GraphEditor/TaskManager.qml:
```diff
@@ -51,222 +62,430 @@ Item {
+                    spacing: 2
+
+                    MaterialToolButton {
+                        ToolTip.text: "Pause job"
```
```suggestion
                        ToolTip.text: "Pause Job"
```

cbentejac commented on meshroom/ui/qml/GraphEditor/TaskManager.qml:
```diff
@@ -51,222 +62,430 @@ Item {
+                    }
+
+                    MaterialToolButton {
+                        ToolTip.text: "Skip task"
```
```suggestion
                        ToolTip.text: "Skip Task"
```

cbentejac commented on meshroom/ui/qml/GraphEditor/TaskManager.qml:
```diff
@@ -51,222 +62,430 @@ Item {
 
-            z: 2
+                    MaterialToolButton {
+                        ToolTip.text: "Stop task"
```
```suggestion
                        ToolTip.text: "Stop Task"
```

cbentejac commented on meshroom/ui/qml/GraphEditor/TaskManager.qml:
```diff
@@ -51,222 +62,430 @@ Item {
-                background: Rectangle {
-                    color: headBgColor
+                    MaterialToolButton {
+                        ToolTip.text: "Restart task"
```
```suggestion
                        ToolTip.text: "Restart Task"
```

cbentejac commented on meshroom/ui/qml/GraphEditor/TaskManager.qml:
```diff
@@ -51,222 +62,430 @@ Item {
+                    }
+
+                    MaterialToolButton {
+                        ToolTip.text: "Resume job"
```
```suggestion
                        ToolTip.text: "Resume Job"
```

cbentejac commented on meshroom/ui/qml/GraphEditor/TaskManager.qml:
```diff
@@ -51,222 +62,430 @@ Item {
+                    }
+
+                    MaterialToolButton {
+                        ToolTip.text: "Interrupt job"
```
```suggestion
                        ToolTip.text: "Interrupt Job"
```

cbentejac commented on meshroom/ui/qml/GraphEditor/TaskManager.qml:
```diff
@@ -51,222 +62,430 @@ Item {
+                    }
+                    
+                    MaterialToolButton {
+                        ToolTip.text: "Restart all error tasks"
```
```suggestion
                        ToolTip.text: "Restart All Error Tasks"
```

cbentejac commented on meshroom/ui/qml/GraphEditor/TaskManager.qml:
```diff
@@ -51,222 +62,430 @@ Item {
+                        ToolTip.text: "Restart all error tasks"
+                        Layout.alignment: Qt.AlignHCenter
+                        enabled: root.uigraph.selectedNode !== null
+                        text: MaterialIcons.replay_circle_filled  // MaterialIcons.restart_alt
```
```suggestion
                        text: MaterialIcons.replay_circle_filled
```

cbentejac reviewed (changes requested):
Additional notes:

- Let's say I compute a graph (for example, the photogrammetry template) in a given location. All my nodes have been successfully computed, the statuses are all set to "SUCCESS". Now let's say I do a "save as" on the graph I just computed and save it to a new location. The cache changes, so we expect all the nodes' status to be reset. For nodes that have dynamic chunks, the display of the status is not refreshed, and we end up with nodes that appear as computed, although they have lost all their status files.
Below is a screenshot of the photogrammetry graph that was computed and saved in a new location (the first branch was computed, the second was added from the template for the sake of comparison):
<img width="1045" height="278" alt="Image" src="https://github.com/user-attachments/assets/32a74a31-1e6d-47d0-b364-b8963575df31" />

- There are some unpredictable behaviours when performing actions that are allowed but probably shouldn't be. An example would be clicking on "stop task" for a task that is already done computing and is in "SUCCESS" state, while nodes are being computed later on in the graph. The UI allows to perform it, there is an info message saying that the task has been successfully stopped, and the node's status is updated to "STOPPED" in the task manager and Graph Editor; it does not enable the "resume job" button. This seems to cause the job on the farm to finish computing the current task and then pausing the rest of the job (to be verified).  

- I have noticed on several occasions that when clicking on buttons from the "JOB" tab, the info message that is sent contains the name of a node that is not part of the job at all (if all my submitted tasks are for NodeName_2, I may get a message about NodeName_1, which is not being computed at all).

cbentejac reviewed (changes requested):
Additional notes:

- Let's say I compute a graph (for example, the photogrammetry template) in a given location. All my nodes have been successfully computed, the statuses are all set to "SUCCESS". Now let's say I do a "save as" on the graph I just computed and save it to a new location. The cache changes, so we expect all the nodes' status to be reset. For nodes that have dynamic chunks, the display of the status is not refreshed, and we end up with nodes that appear as computed, although they have lost all their status files.
Below is a screenshot of the photogrammetry graph that was computed and saved in a new location (the first branch was computed, the second was added from the template for the sake of comparison):
<img width="1045" height="278" alt="Image" src="https://github.com/user-attachments/assets/32a74a31-1e6d-47d0-b364-b8963575df31" />

- There are some unpredictable behaviours when performing actions that are allowed but probably shouldn't be. An example would be clicking on "stop task" for a task that is already done computing and is in "SUCCESS" state, while nodes are being computed later on in the graph. The UI allows to perform it, there is an info message saying that the task has been successfully stopped, and the node's status is updated to "STOPPED" in the task manager and Graph Editor; it does not enable the "resume job" button. This seems to cause the job on the farm to finish computing the current task and then pausing the rest of the job (to be verified).  

- I have noticed on several occasions that when clicking on buttons from the "JOB" tab, the info message that is sent contains the name of a node that is not part of the job at all (if all my submitted tasks are for NodeName_2, I may get a message about NodeName_1, which is not being computed at all).

Alxiice committed:
[chunks] Apply typos/cleaning suggestions from @cbentejac

Alxiice committed:
[submitter] Add tools to avoid autoretry on farm

Alxiice committed:
[submitter] Fix interruptJob UI updates

cbentejac committed:
[bin] Update permissions on `meshroom_createChunks`

cbentejac committed:
[core] node: Correctly use custom size for non-parallelized nodes

cbentejac committed:
[bin] Fix typo: Replace occurrences of "infos" with "info"

cbentejac committed:
[core] Fix typo: Replace all occurrences of "infos" with "info"

cbentejac committed:
[core] Linting: Remove all trailing whitespaces

cbentejac committed:
[core] node: Remove references to `packageVersion` in `NodeStatusData`

cbentejac committed:
[core] node: Remove static info from the chunks' status file

`nodeType` and `packageName` are already part of the node's status file,
and there is thus no need to propagate them to the chunks'.

cbentejac committed:
Linting: Remove trailing whitespaces

cbentejac committed:
[core] node: Use explicit keys for chunks' blockSize, fullSize and nbBlocks in status files

cbentejac committed:
[core] node: Detect whether external jobs can be stopped or canceled

`canBeStopped` and `canBeCanceled` only took into account nodes that
were being run locally. As nodes that have been submitted externally
(exclusively on a render farm, not in another instance of Meshroom) can
now be stopped and canceled as well, they are considered in both methods.

cbentejac committed:
[ui] Add `stoppable` state for the `Submit` button

When a job has been submitted, the state of the `Submit` button switches
to `stoppable`. Clicking it in this state interrupts the node's computation
on the farm.

cbentejac committed:
[ui] NodeActions: Add `Retry` button for submitted tasks on error state

cbentejac committed:
[GraphEditor] NodeChunks: Remove specific color for dynamic chunks

cbentejac committed:
[ui] NodeActions: Fix status of `compute` and `submit` in `deletable` state

cbentejac committed:
[GraphEditor] Add "Retry Error Tasks" menu to match NodeActions

cbentejac committed:
[GraphEditor] Add "Interrupt/Cancel Job" menus

cbentejac committed:
[GraphEditor] NodeChunks: Don't add specific display when there's no chunk

cbentejac committed:
[core] node: Add `chunkPlaceholder` property

`chunkPlaceholder` is a list model that is meant to contain a placeholder
chunk for nodes that have dynamic chunks: prior to the chunks' creation,
it is used to reflect the state of the node as a single chunk.

In particular, this allows to reflect accurately the node's status while
it is submitted but the chunks have not been created yet.

The property is a list model as it allows it to be used as a model later
on on the QML side.

cbentejac committed:
[ui] Use chunk placeholder for uncreated dynamic chunks

cbentejac committed:
[core] graph: Trigger `onGraphUpdated` slot when chunks change

`onGraphUpdated` updates the models for chunks, including the placeholder,
which allows to have an up-to-date status for all the existing chunks as
well as the placeholder one when a node's number of chunks hasn't been
determined yet.

cbentejac committed:
[ui] Application: Update state of the global "Submit" icon when needed

Alxiice commented on meshroom/core/desc/computation.py:
```diff
@@ -1,13 +1,15 @@
     NORMAL = 1
     INTENSIVE = 2
+    EXTREME = 3
+    SCRIPT=-1
```
We can put it before NONE if it's better but the reason is that's levels are going from less system-inensive to more intensivity.

By the way jobs with script level are not starting now in our tractor instance we should see with Thibault why is that

Alxiice commented on meshroom/core/taskManager.py:
```diff
@@ -426,17 +534,20 @@ def submit(self, graph, submitter=None, toNodes=None, submitLabel="{projectName}
                 for node in nodesToProcess:
-                    node.destroyed.connect(lambda obj=None, name=node.name: self.onNodeDestroyed(obj, name))
-                    node.initStatusOnSubmit()  # update node status
+                    # TODO : Notify the node that there was an issue on submit
```
to do

Alxiice commented on meshroom/core/graph.py:
```diff
@@ -1730,11 +1755,19 @@ def submitGraph(graph, submitter, toNodes=None, submitLabel="{projectName}"):
+        else:
             for node in nodesToProcess:
-                node.initStatusOnSubmit()  # update node status
+                # TODO : Notify the node that there was an issue on submit
```
to do

Alxiice commented on meshroom/core/node.py:
```diff
@@ -1644,26 +1930,28 @@ def updateDuplicates(self, nodesPerUid):
-        return True
+        """ Check if the node is submitted by the current session
+        TODO: Not used -> depreciate ?
+        """
```
```suggestion
```

Alxiice commented on meshroom/core/node.py:
```diff
@@ -1644,26 +1930,28 @@ def updateDuplicates(self, nodesPerUid):
+            if chunk._status.computeSessionUid != meshroom.core.sessionUid:
                 return False
         return True
 
```
```suggestion

```

cbentejac committed:
[core] node: Correctly use `fullSize` instead of `nbChunks` to set chunks from cache

cbentejac committed:
`.git-blame-ignore-revs`: Add linting commits

Alxiice committed:
Correct typos

Alxiice committed:
[node] remove methods statusInThisSession & submitterStatusInThisSession

Alxiice committed:
[graph] Fix a small issue in compareFilesTimes

cbentejac committed:
[core] node: Reset chunks before updating the node's size
